Mathlib/Analysis/SpecialFunctions/OrdinaryHypergeometric.lean

```diff
@@ -0,0 +1,282 @@
+theorem ordinaryHypergeometricSeries_apply_eq' (x : 𝔸) :
+    (fun n => ordinaryHypergeometricSeries 𝔸 a b c n fun _ => x) =
+    fun n => ((n !⁻¹ : 𝕂) * (ascPochhammer 𝕂 n).eval a * (ascPochhammer 𝕂 n).eval b *
+    ((ascPochhammer 𝕂 n).eval c)⁻¹ ) • x ^ n := by
+  simp [ordinaryHypergeometricSeries]
+
+theorem ordinaryHypergeometric_sum_eq (x : 𝔸) : (ordinaryHypergeometricSeries 𝔸 a b c).sum x =
+    ∑' n : ℕ, ((n !⁻¹ : 𝕂) * (ascPochhammer 𝕂 n).eval a * (ascPochhammer 𝕂 n).eval b *
+    ((ascPochhammer 𝕂 n).eval c)⁻¹ ) • x ^ n :=
+  tsum_congr fun n => ordinaryHypergeometricSeries_apply_eq a b c x n
+
+theorem ordinaryHypergeometric_eq_tsum : ₂F₁ a b c =
+    fun (x : 𝔸) => ∑' n : ℕ, ((n !⁻¹ : 𝕂) * (ascPochhammer 𝕂 n).eval a *
+    (ascPochhammer 𝕂 n).eval b * ((ascPochhammer 𝕂 n).eval c)⁻¹ ) • x ^ n :=
+  funext (ordinaryHypergeometric_sum_eq a b c)
+
+theorem ordinaryHypergeometricSeries_apply_zero (n : ℕ) :
+    (ordinaryHypergeometricSeries 𝔸 a b c n fun _ => (0:𝔸)) =
+    Pi.single (f := fun _ => 𝔸) 0 1 n := by
+  rw [ordinaryHypergeometricSeries_apply_eq]
+  cases n <;> simp
+
+@[simp]
+theorem ordinaryHypergeometric_zero : ₂F₁ a b c (0 : 𝔸) = 1 := by
+  simp [ordinaryHypergeometric_eq_tsum, ← ordinaryHypergeometricSeries_apply_eq,
+    ordinaryHypergeometricSeries_apply_zero]
+
+@[simp]
+theorem ordinaryHypergeometric_op [T2Space 𝔸] (x : 𝔸) :
+    ₂F₁ a b c (MulOpposite.op x) = MulOpposite.op (₂F₁ a b c x) := by
+  simp [ordinaryHypergeometric, ordinaryHypergeometric_sum_eq, ← MulOpposite.op_pow,
+    ← MulOpposite.op_smul, tsum_op]
+
+@[simp]
+theorem ordinaryHypergeometric_unop [T2Space 𝔸] (x : 𝔸ᵐᵒᵖ) :
+    ₂F₁ a b c (MulOpposite.unop x) = MulOpposite.unop (₂F₁ a b c x) := by
+  simp [ordinaryHypergeometric, ordinaryHypergeometric_sum_eq, ← MulOpposite.unop_pow,
+     ← MulOpposite.unop_smul, tsum_unop]
+
+theorem ordinaryHypergeometricSeries_symm :
+    ordinaryHypergeometricSeries 𝔸 a b c = ordinaryHypergeometricSeries 𝔸 b a c := by
+  ext
+  simp [ordinaryHypergeometricSeries]
+  group
+
+/-- If any parameter to the series is a sufficiently large nonpositive integer, then the series
+term is zero. -/
+lemma ordinaryHypergeometricSeries_eq_zero_of_neg_nat {n k : ℕ}
+    (habc : k = -a ∨ k = -b ∨ k = -c) (hk : k < n) :
+    ordinaryHypergeometricSeries 𝔸 a b c n = 0 := by
+  rw [ordinaryHypergeometricSeries]
+  rcases habc with h | h | h
+  all_goals
+    ext
+    simp [(ascPochhammer_eval_eq_zero_iff n _).2 ⟨k, h, hk⟩]
+
+end Field
+
+section RCLike
+
+open Asymptotics Filter Real Set Nat
+
+variable {𝕂 : Type*} (𝔸 𝔹 : Type*) [RCLike 𝕂] [NormedDivisionRing 𝔸] [NormedAlgebra 𝕂 𝔸]
+  (a b c : 𝕂)
+
+theorem ordinaryHypergeometric_radius_top_of_neg_nat₁ {ak : ℕ} (ha : ak = -a) :
+    (ordinaryHypergeometricSeries 𝔸 a b c).radius = ⊤ := by
+  apply FormalMultilinearSeries.radius_eq_top_of_forall_image_add_eq_zero _ (1 + ak)
+  exact fun _ ↦ ordinaryHypergeometricSeries_eq_zero_of_neg_nat a b c (Or.inl ha) (by linarith)
```

Comment:
As mentioned elsewhere, it's almost never desirable to have a hypothesis containing an equality. So we should spell this:
```suggestion
theorem ordinaryHypergeometric_radius_top_of_neg_nat₁ {k : ℕ} :
    (ordinaryHypergeometricSeries 𝔸 (-(k : 𝕂)) b c).radius = ⊤ := by
  refine FormalMultilinearSeries.radius_eq_top_of_forall_image_add_eq_zero _ (1 + k) fun n ↦ ?_
  exact ordinaryHypergeometricSeries_eq_zero_of_neg_nat (-(k : 𝕂)) b c
    (Or.inl (neg_neg _).symm) (by omega)
```
and likewise below